### PR TITLE
Restructure around Python 3.10 Pattern Matching

### PR DESCRIPTION
**Pull request by dpassen**
- Add type hints
- Lint and format with ruff

### CODE REVIEW

**Comment by dpassen** [comment]
There's a lot here. Need to figure out coveralls again, and also, put a note pointing users of older Python version to the currently published version.

**Comment by cladden** [comment]
3.10 and type hint support would be so awesome!

**Comment by dpassen** [comment]
@cbefus

**Review by cbefus-indeed — DISMISSED** [review]
Some really interesting changes here.  LGTM.

**Review by cbefus — DISMISSED** [review]
Ignore that other guy - this is the real me.  LGTM

**Comment by cbefus** [comment]
> 

Need to update releases.
Probably need to update the Readme as well?

**Comment by dpassen** [comment]
I may need a little help from you in that regard

**Comment by dpassen** [comment]
Found an issue, needs more thought

**Comment by cbefus** [comment]
Let me know when you are ready for a re-review and then I will figure out deploy.

File: optional/optional.py [hunk 1]
```diff
@@ -1,12 +1,40 @@
...
+
+
+class Optional:
+    """Entrypoint of optional.py
+
+    Provides static methods to help intelligently create
+    both :class:`Something` and :class:`Nothing` values"""
+
+    @staticmethod
+    @overload
```
Comment (cbefus) [inline optional/optional.py:19]: I assume these overloads somehow offload to the single implementation?

Comment (dpassen) [reply optional/optional.py:19]: Yeah. It's for the typing. Consider that `of` technically takes a `(T | None)` and returns an `Option[T]`, but that type signature doesn't make it clear that `None -> Nothing` and `T -> Option[T]`. These overloads encode that information

**Comment by dpassen** [comment]
I can make time to mob this together

**Comment by dpassen** [comment]
@cbefus

**Review by cbefus — APPROVED** [review]

**Comment by dpassen** [comment]
@cbefus Oh heck yeah. Hope you remember how to publish!

**Pull request merged**